[change:tests] Switch to ChannelsLiveServerTestCase #904

dee077 opened the pull request:
Replaced StaticLiveServerTestCase with ChannelsLiveServerTestCase to resolve JS errors caused by failed websocket connections.

Fixes #904

## Checklist

- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [x] I have manually tested the changes proposed in this pull request.

## Reference to Existing Issue

Closes #904.

## Description of Changes

- Changed StaticLiveServerTestCase to ChannelsLiveServerTestCase.
- Updated to Django 5.2.0 in ci.
- Removed --parallel for running test in ci.

### Issues

- ChannelsLiveServerTestCase does not support Django 5.2 yet.
- For running test --parallel does not work.

nemesifier reviewed (commented):
Looks on the right track to me.

Question: how can we be sure it's working as expected and there's no JS errors?
Wouldn't it make sense to add at least one test for this?

CI is not passing on Django 5.2, why is that?

dee077 commented on tests/openwisp2/settings.py:
```diff
@@ -77,7 +77,13 @@
+        'ENGINE': 'django.db.backends.sqlite3',
+        'NAME': 'openwisp_utils.db',
+        'TEST': {
+            'NAME': 'openwisp_utils_test.db',
```
> Why is this needed?

Without this, it was causing this error
```
raise ImproperlyConfigured(
django.core.exceptions.ImproperlyConfigured: ChannelLiveServerTestCase can not be used with in memory databases
```
As the ASGI server requires a physical database

dee077 commented:
> CI is not passing on Django 5.2, why is that?

There is an open issue in django channels for ChannelsLiveServerTestCase is broken in django 5.2

nemesifier commented on setup.py:
```diff
@@ -47,7 +47,9 @@
-            "channels[daphne]~=4.2.0",
+            "channels[daphne]@"
+            "git+https://github.com/openwisp/channels.git@"
+            "issues/2148-ChannelsLiveServerTestCase-Django52-compatible#egg=channels[daphne]",
```
We can update this now as the PR has been merged.

nemesifier commented on .github/workflows/ci.yml:
```diff
@@ -86,7 +86,7 @@ jobs:
       if: ${{ !cancelled() && steps.deps.conclusion == 'success' }}
       run: |
-        coverage run runtests.py --parallel
+        coverage run runtests.py
```
We need to handle this.

nemesifier commented on .github/workflows/ci.yml:
```diff
@@ -79,14 +79,16 @@ jobs:
         pip install ${{ matrix.django-version }}
         sudo npm install -g prettier
+        pip uninstall channels channels-redis daphne -y
+        pip install -e ".[channels]"
```
Shouldn't be needed anymore.

dee077 commented:
## Updates:
- Changed `ChannelsLiveServerTestCase` from `StaticLiveServerTestCase`
- Added tag `selenium tests` to all selenium tests to exclude them from running in parallel.
- Updated CI to run unit and selenium tests separately.

nemesifier reviewed (approved):
LGTM! @pandafy what do you think?

pandafy commented on tests/openwisp2/asgi.py:
```diff
@@ -0,0 +1,6 @@
+
+from django.core.asgi import get_asgi_application
+
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "openwisp2.settings")
```
Why do we need this here? I don't see it in openwisp-controller and openwisp-notifications.

pandafy reviewed (approved):
LGTM! I have one question, the rest looks good!
